RecoEgamma/EgammaTools/plugins/HGCalElectronIDValueMapProducer.cc

```diff
@@ -0,0 +1,238 @@
+/**\class HGCalElectronIDValueMapProducer HGCalElectronIDValueMapProducer.cc RecoEgamma/HGCalElectronIDValueMapProducer/plugins/HGCalElectronIDValueMapProducer.cc
+
+ Description: [one line class summary]
+
+ Implementation:
+    [Notes on implementation]
+*/
+//
+// Original Author:  Nicholas Charles Smith
+//      Created:  Wed, 05 Apr 2017 12:17:43 GMT
+//
+//
+
+
+// system include files
+#include <memory>
+
+// user include files
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+
+#include "DataFormats/Common/interface/ValueMap.h"
+
+#include "DataFormats/EgammaCandidates/interface/GsfElectron.h"
+#include "DataFormats/EgammaCandidates/interface/GsfElectronFwd.h"
+#include "DataFormats/GsfTrackReco/interface/GsfTrack.h"
+
+#include "RecoEgamma/EgammaTools/interface/ElectronIDHelper.h"
+#include "RecoEgamma/EgammaTools/interface/LongDeps.h"
+
+class HGCalElectronIDValueMapProducer : public edm::stream::EDProducer<> {
+  public:
+    explicit HGCalElectronIDValueMapProducer(const edm::ParameterSet&);
+    ~HGCalElectronIDValueMapProducer() override;
+
+    static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  private:
+    void beginStream(edm::StreamID) override;
+    void produce(edm::Event&, const edm::EventSetup&) override;
+    void endStream() override;
+
+    // ----------member data ---------------------------
+    edm::EDGetTokenT<edm::View<reco::GsfElectron>> ElectronsToken_;
+    float radius_;
+    std::map<const std::string, std::vector<float>> maps_;
+
+    std::unique_ptr<ElectronIDHelper> eIDHelper_;
+};
+
+HGCalElectronIDValueMapProducer::HGCalElectronIDValueMapProducer(const edm::ParameterSet& iConfig) :
+  ElectronsToken_(consumes<edm::View<reco::GsfElectron>>(iConfig.getParameter<edm::InputTag>("electrons"))),
+  radius_(iConfig.getParameter<double>("pcaRadius"))
+{
+  // All the ValueMap names to output are defined in the python config
+  // so that potential consumers can configure themselves in a simple manner
+  for(auto key : iConfig.getParameter<std::vector<std::string>>("variables")) {
+    maps_[key] = {};
+    produces<edm::ValueMap<float>>(key);
+  }
+
+  eIDHelper_.reset(new ElectronIDHelper(iConfig, consumesCollector()));
+}
+
+
+HGCalElectronIDValueMapProducer::~HGCalElectronIDValueMapProducer()
+{
+}
+
+
+// ------------ method called to produce the data  ------------
+void
+HGCalElectronIDValueMapProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
+{
+  using namespace edm;
+
+  Handle<edm::View<reco::GsfElectron>> ElectronsH;
+  iEvent.getByToken(ElectronsToken_, ElectronsH);
+
+  const size_t prevMapSize = maps_.size();
+
+  // Clear previous map
+  for(auto&& kv : maps_) kv.second.clear();
```

Comment:
Since these vectors are filled for every electron, one could also `reserve()` here based on the electron collection size.

Reply:
It's true that `clear()` won't remove the existing memory reservation. What is the average/max size of the electrons collection in 200PU? If there can be large fluctuations, it may still be worthwhile to prevent a lot of allocations from outlier events.